fvm/environment/unsafe_random_generator.go

```diff
@@ -59,31 +60,47 @@ func (gen ParseRestrictedUnsafeRandomGenerator) UnsafeRandom() (
 func NewUnsafeRandomGenerator(
 	tracer tracing.TracerSpan,
 	blockHeader *flow.Header,
+	txnIndex uint32,
 ) UnsafeRandomGenerator {
 	gen := &unsafeRandomGenerator{
 		tracer:      tracer,
 		blockHeader: blockHeader,
+		txnIndex:    txnIndex,
 	}
 
 	return gen
 }
 
-// This function abstracts building the PRG seed from the entropy source `randomSource`.
-// It does not make assumptions about the quality of the source, nor about
-// its length (the source could be a fingerprint of entity, an ID of an entity,
+// This function abstracts building the PRG seed from the entropy source
+// `randomSource` and `salt`. It does not make assumptions about the quality
+// of the source, nor about its length (the source could be a fingerprint of
+// entity, an ID of an entity, a beacon signature..)
 //
-//	a beacon signature..)
-//
-// It therefore uses a mechansim to extract the source entropy and expand it into
-// the required `seedLen` bytes (this can be a KDF, a MAC, a hash with extended-length output..)
-func seedFromEntropySource(randomSource []byte, seedLen int) ([]byte, error) {
+// It therefore uses a mechansim to extract the source entropy and expand it
+// into the required `seedLen` bytes (this can be a KDF, a MAC, a hash with
+// extended-length output..)
+func seedFromEntropySource(
```

Comment:
nit: this implementation is a bit too complicated. Both input and Output sizes are constant, but we pass slices around, essentially moving all size checks to runtime.  Also this complicates codebase a lot, given that we now need an expansion function instead of a dump hmac.

Reply:
I can understand why the extra helper looks like a code complication, but from a cryptographic point of view, it is a separated step from the rest of the logic and its implementation details can be abstracted away from the rest. This can allow updating the extract-and-expand logic (whether it is a dump hmac, hash, or a generic XOF..) without touching the core logic.